### ISSUE TITLE
"kubectl logs --follow" has not printed new logs after log rotation. 

### ISSUE THREAD

**Issue by s-kawamura-w664** [kubernetes/kubernetes#108678]
### What happened?

When running `kubectl logs pod-name --follow`, new logs are not output to the logs command after the log file has grown and rotated.

### What did you expect to happen?

Even after log rotation, new logs are output to `kubectl logs pod-name --follow`.

### How can we reproduce it (as minimally and precisely as possible)?

The following steps will reproduce it.

1. Deploy a Pod that continues to output to stdout.
2. Run `kubectl logs pod-name --follow`.
3. Wait a while and the log file will be rotated. The waiting time depends on the environment.
4. New logs will not be output to logs command.

The sample pod is following.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-print
spec:
  containers:
  - name: centos7
    image: centos:7
    imagePullPolicy: Never
    command: ["sh", "-c"]
    args:
    - cnt=0;
      msg="";
      while [ $cnt -lt 5000 ];
      do
        msg="${msg} ${cnt} :loop ---------------------------";
        cnt=`expr $cnt + 1`;
      done;
      loopcnt=0;
      while true; do
        echo $msg;
        date;
        echo "loop= $loopcnt";
        loopcnt=`expr $loopcnt + 1`;
        sleep 2;
      done;
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-03-06T21:32:53Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-03-06T21:32:53Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
not cloud
</details>

### OS version

<details>

```console
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 21.10"
NAME="Ubuntu"
VERSION_ID="21.10"
VERSION="21.10 (Impish Indri)"
VERSION_CODENAME=impish
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=impish
$ uname -a
Linux kind-control-plane 4.15.0-166-generic #174-Ubuntu SMP Wed Dec 8 19:07:44 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by s-kawamura-w664** [comment]
I have investigated this issue.
I found that when the problem occurs, it always goes through the same place and then there is a problem.

https://github.com/kubernetes/kubernetes/blob/897f2f78a5c7ff659dfce23f53dd36b9261a23c9/pkg/kubelet/kuberuntime/logs/logs.go#L351-L365

When a log file is renamed by rotation, `recreated` becomes true, but immediately after the renaming, `os.Open(path)` will result in a `Not Exist` error because the new log file has not yet been created.

Then it will re-loop by continue, but since `recreated` will not be true in the next loop, it will never enter this process again.
As a result, the log after log rotation will not be read.
